### ISSUE TITLE
Auto logout by idle timeout

### ISSUE THREAD

**Issue by mrdokenny** [bitwarden/browser#173]
Perhaps add an option to have the extension detect when the user is idle before counting down the timer to log out. Thanks for your hard work!

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/idle

**Comment by kspearrin** [comment]
Good idea. Didn't know about this API.

**Comment by kspearrin** [comment]
added "On Locked" option for next release: https://github.com/bitwarden/browser/commit/96be6466412dbfac8b6a0f24fd10c1b2c4c0138e

**Comment by mrdokenny** [comment]
Nice that was fast.

**Comment by mrdokenny** [comment]
I just tried it out by building it. It's a good start but I think it should be like a checkmark under lock items named "Check if idle". Then if idle ->count down the timer depending on what the user sets.

Use case:
Settings: {Check if idle: true, timer: 5 minutes}
- User logs into the extension
- While user is actively using the browser the extension doesn't lock
- Only when the user leaves the computer and browser state is idle, then the timer will tick down
- When the timer counts down to 0, then it will lock
- Say 1 minute (or whatever you want or provide options) to check if idle + 5 minute timer = 6 minutes to lock the extension after idle

**Comment by kspearrin** [comment]
Kenny,
I still plan to add other options like you mentioned. This was just a quick win I got out of checking out the new API that I could get in with this release.

One problem that we will have to solve to get the effect you are looking for is what is the definition of "idle". The API doesn't just tell you when the user is idle, you have to tell it what idle means by specifying a number of seconds on inactivity. See https://developer.chrome.com/apps/idle#method-setDetectionInterval

What what is idle?

**Comment by mrdokenny** [comment]
From what I remember, how Lastpass does it is they have a separate drop down menu for idle timer. So the user can choose how long before the browser determines when it is "idle".

**Comment by kspearrin** [comment]
Seems like a lot of configuration. Need to think about this more.

**Comment by mrdokenny** [comment]
Another option like I said before is to have a checkbox "Check if idle" then you can decide how long "idle" would be, but it would be handy to have another drop down menu.

**Comment by jungle-boogie** [comment]
I support this request. On accident I opened a request at core, but this is really a browser thing.

I think a simple auto logout of 30/60/90 minutes wouldn't be too complicated for users.

https://github.com/bitwarden/core/issues/146

**Comment by ToNIX-** [comment]
I'd also like to see a logout after X mins of idle instead of a fixed time.

**Comment by bricep** [comment]
I agree that an idle lockout is pretty much a necessity. It looks like the idle API could be used here and is compatible across most browsers:

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/idle

**Comment by kustodian** [comment]
I really like how LastPass did it. It only works with *idle times* and there are two separate idle times:
* while browser is running
* while the browser is closed

Here is how it looks like:
![image](https://user-images.githubusercontent.com/1949229/34515316-e0172ab0-f071-11e7-8dbf-3794e7b050ed.png)

This means that you can configure to lock the vault after N minutes of innactivity or after a close (restart) of a browser. Bitwarden only has a timeout to configure either of the two, and you need both for security.

I would also suggest better sane defaults for Lock, because leaving Bitwarden unlocked all the time (if I remember correctly that is the default) is no the best option for the majority of users.

**Comment by ajayyy** [comment]
https://github.com/bitwarden/browser/pull/1159
